src/colvarbias.cpp

```diff
@@ -268,6 +267,16 @@ int colvarbias::bin_count(int bin_index)
   cvm::error("Error: bin_count() not implemented.\n");
   return COLVARS_NOT_IMPLEMENTED;
 }
+cvm::real colvarbias::bin_gradient(int bin_index)
+{
+  cvm::error("Error: bin_gradient() not implemented.\n");
+  return COLVARS_NOT_IMPLEMENTED;
+}
+cvm::real colvarbias::delta_free_energy(cvm::real x)
```

Comment:
These two functions are by design incompatible with multiple variables, or non scalar variables.  Rather than overloading functions later, which will be confusing, I'd recommend using `colvarvalue` instead of `cvm::real` and `std::vector<int>` instead of just `int`.

Reply:
> Can you use the existing energy command to return the bias's energy, i.e. the negative of what you return as freeenergy? This would allow replica exchange scripts to be more uniform between biases.

As you noted below, I sometimes need to be able to get the bias energy at positions other than current. It does seem reasonable that colvarbias_abf should set bias_energy like the other biases. I could then, as you suggested below, just use "cv bias abf1 energy" to get the current energy. It looks like I could just set bias_energy in colvarbias_abf::update. If I do this, will the ABF energy be included in the energies reported by NAMD?

> If only 1D ABF is tested, I strongly recommend making this feature incompatible with multi-dimensional ABF.

I'm unsure how to check if the ABF is multidimensional.

> In 2D and 3D ABF, you could rely on the existing integrated gradients, especially if they are only necessary at sparse times. The fullSamples question is less pressing then, because noisy gradient estimates get smoothed out by the Poisson integration process.

Where do I get these integrated gradients?

I'd also really like to make this work for eABF, but I don't quite understand how to do this yet. I'm a little confused by "cv[i]->actual_value()" and "cv[i]->value()".

Reply:
> I'd recommend using colvarvalue instead of cvm::real and std::vector<int> instead of just int

I will try this.

Reply:
About `bias_energy`, that's correct: if ABF starts setting it based on the integrated gradients (1D, 2D or 3D) NAMD will report that energy as well.  Please coordinate with @jhenin on this as well as how to use the multi-dim integrators and eABF.